dev/extra-ml-requirements.txt

```diff
@@ -42,3 +42,14 @@ pyspark
 shap
 # Required by mlflow.paddle
 paddlepaddle
+# Required by mlflow.prophet
+Cython
+convertdate
+LunarCalendar
+holidays
+tqdm
+cmdstanpy
+setuptools-git
+python-dateutil
+pystan~=2.19.1.1
```

Comment:
```suggestion
```

Can we remove these packages since they are installed with `prophet`?

Reply:
As mentioned above, the build process for Prophet will fail if these aren't on the path prior to pystan and prophet being installed. They're using a rather non-standard way for installing dependencies in their lib and not having them present will cause the pip install of prophet and pystan to fail.

Reply:
let's add an inline comment here to clarify why we're installing all of these dependencies for future readers

Reply:
added comment

Reply:
Got it. If so, we need to install these before installing `prophet`, right?

```
pip install Cython convertdate ...
pip install -r dev/extra-ml-requirements.txt
```

Reply:
Was hoping that we could pass ordered packages to the same `pip install` command. Looks like this fails with ```DEPRECATION: prophet was installed using the legacy 'setup.py install' method, because a wheel could not be built for it. A possible replacement is to fix the wheel build issue reported above. You can find discussion regarding this at https://github.com/pypa/pip/issues/8368.
Successfully installed Cython-0.29.24 LunarCalendar-0.0.9 absl-py-0.13.0 aiohttp-3.7.4.post0 applicationinsights-0.11.10 astor-0.8.1 astunparse-1.6.3 async-timeout-3.0.1 attrs-21.2.0 azure-identity-1.6.1 azure-mgmt-resource-8.0.1 azureml-automl-core-1.2.0 azureml-core-1.2.0.post3 azureml-dataprep-1.3.6 azureml-dataprep-native-14.2.1 azureml-pipeline-1.2.0 azureml-pipeline-core-1.2.0 azureml-pipeline-steps-1.2.0 azureml-sdk-1.2.0 azureml-telemetry-1.2.0 azureml-train-1.2.0 azureml-train-automl-client-1.2.0 azureml-train-core-1.2.0 azureml-train-restclients-hyperdrive-1.2.0 beautifulsoup4-4.10.0 blis-0.7.4 bottleneck-1.3.2 cached-property-1.5.2 catalogue-2.0.6 catboost-0.26.1 chardet-4.0.0 clang-5.0 click-7.1.2 cmdstanpy-0.9.68 colorama-0.4.4 contextvars-2.4 convertdate-2.3.2 cymem-2.0.5 decorator-5.0.9 distro-1.6.0 dotnetcore2-2.1.21 ephem-4.0.0.2 fastai-1.0.60 fastprogress-1.0.0 flake8-3.7.9 fsspec-2021.8.1 fusepy-3.0.1 future-0.18.2 gast-0.4.0 google-auth-1.35.0 google-auth-oauthlib-0.4.6 google-pasta-0.2.0 graphviz-0.8.4 grpcio-1.40.0 h2o-3.32.1.7 h5py-3.1.0 hijri-converter-2.2.1 holidays-0.11.2 idna-ssl-1.1.0 immutables-0.16 keras-2.6.0 keras-preprocessing-1.1.2 korean-lunar-calendar-0.2.1 lightgbm-3.2.1 llvmlite-0.36.0 markdown-3.3.4 mccabe-0.6.1 msal-1.14.0 msal-extensions-0.3.0 multidict-5.1.0 murmurhash-1.0.5 mxnet-1.8.0.post0 numba-0.53.1 numexpr-2.7.3 nvidia-ml-py3-7.352.0 onnxruntime-1.8.1 opt-einsum-3.3.0 paddlepaddle-2.1.2 pathy-0.6.0 patsy-0.5.1 portalocker-1.7.1 preshed-3.0.5 prophet-1.0.1 py4j-0.10.9 pyDeprecate-0.3.1 pycodestyle-2.5.0 pydantic-1.8.2 pyflakes-2.1.1 pymeeus-0.5.11 pyspark-3.1.2 pystan-2.19.1.1 pytorch-lightning-1.4.5 ruamel.yaml-0.15.89 setuptools-git-1.2 shap-0.39.0 six-1.15.0 slicer-0.0.7 smart-open-5.2.1 soupsieve-2.2.1 spacy-3.1.2 spacy-legacy-3.0.8 srsly-2.4.1 statsmodels-0.12.2 tensorboard-2.6.0 tensorboard-data-server-0.6.1 tensorboard-plugin-wit-1.8.0 tensorflow-2.6.0 tensorflow-estimator-2.6.0 termcolor-1.1.0 thinc-8.0.10 torch-1.9.0 torchmetrics-0.5.1 torchvision-0.10.0 tqdm-4.62.2 typer-0.3.2 typing-extensions-3.7.4.3 ujson-4.1.0 wasabi-0.8.2 wrapt-1.12.1 xgboost-1.4.2 yarl-1.6.3```

However, prophet installation does ultimately succeed. Thoughts @BenWilson2 ?

Reply:
https://github.com/mlflow/mlflow/pull/4773/checks?check_run_id=3550111462#step:5:1038

```
    Running setup.py install for prophet: started
    Running setup.py install for prophet: still running...
    Running setup.py install for prophet: finished with status 'done'
  DEPRECATION: prophet was installed using the legacy 'setup.py install' method, because a wheel could not be built for it. A possible replacement is to fix the wheel build issue reported above. You can find discussion regarding this at https://github.com/pypa/pip/issues/8368.
```

The error log above indicates that pip attempted to install `prophet` by running `setup.py install`. I guess prophet's dependencies are already installed when running `setup.py install`.

Reply:
The installation order is determined by `get_installation_order`:

https://github.com/pypa/pip/blob/5ab7c124db917625e55693773bcc224793d7ceed/src/pip/_internal/resolution/resolvelib/resolver.py#L185

```python
    def get_installation_order(
        self, req_set: RequirementSet
    ) -> List[InstallRequirement]:
        """Get order for installation of requirements in RequirementSet.

        The returned list contains a requirement before another that depends on
        it. This helps ensure that the environment is kept consistent as they
        get installed one-by-one.
        ...
        """
```

Reply:
`prophet` can be installed by simply running `pip install prophet`:

```dockerfile
# Dockerfile

FROM python:3.7

RUN pip install prophet
RUN python -c "import prophet"

# `docker build .` succeeds
```

Reply:
Interesting. IIRC we had trouble with just including `prophet` in the list of dependencies. Conda environment creation failed. @harupy @BenWilson2 can you try this out with just `prophet` and confirm that it fails or succeeds?

Reply:
```dockerfile
FROM continuumio/miniconda3

RUN echo '\
name: prophet\n\
channels:\n\
  - conda-forge\n\
dependencies:\n\
  - python=3.6\n\
  - pip\n\
  - pip:\n\
      - prophet\n\
' > conda.yaml
RUN cat conda.yaml
RUN conda env create --file conda.yaml
```

failed with:

```
  ModuleNotFoundError: No module named 'numpy'
  ----------------------------------------
  ERROR: Failed building wheel for prophet
    ERROR: Command errored out with exit status 1:
     command: /opt/conda/envs/prophet/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-spruk13i/prophet_286c94955d8f495093472c3bf4fb6f23/setup.py'"'"'; __file__='"'"'/tmp/pip-install-spruk13i/prophet_286c94955d8f495093472c3bf4fb6f23/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-x79jrygt/install-record.txt --single-version-externally-managed --compile --install-headers /opt/conda/envs/prophet/include/python3.6m/prophet
         cwd: /tmp/pip-install-spruk13i/prophet_286c94955d8f495093472c3bf4fb6f23/
    Complete output (10 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib
    creating build/lib/prophet
    creating build/lib/prophet/stan_model
    Importing plotly failed. Interactive plots will not work.
    INFO:pystan:COMPILING THE C++ CODE FOR MODEL anon_model_dfdaf2b8ece8a02eb11f050ec701c0ec NOW.
    error: command 'gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /opt/conda/envs/prophet/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-spruk13i/prophet_286c94955d8f495093472c3bf4fb6f23/setup.py'"'"'; __file__='"'"'/tmp/pip-install-spruk13i/prophet_286c94955d8f495093472c3bf4fb6f23/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-x79jrygt/install-record.txt --single-version-externally-managed --compile --install-headers /opt/conda/envs/prophet/include/python3.6m/prophet Check the logs for full command output.


CondaEnvException: Pip failed
```

Reply:
@harupy Any idea how the current approach seems to work? Does gcc get installed?

Reply:
This doesn't work either:

```python
FROM continuumio/miniconda3

RUN echo '\
name: prophet\n\
channels:\n\
  - conda-forge\n\
dependencies:\n\
  - python=3.6\n\
  - pip\n\
  - pip:\n\
      - Cython\n\
      - convertdate\n\
      - LunarCalendar\n\
      - holidays\n\
      - tqdm\n\
      - cmdstanpy\n\
      - setuptools-git\n\
      - python-dateutil\n\
      - pystan~=2.19.1.1\n\
      - prophet\n\
' > conda.yaml

RUN conda env create --file conda.yaml
```

Reply:
This works:

```dockerfile
FROM continuumio/miniconda3

RUN echo '\
name: prophet\n\
channels:\n\
  - conda-forge\n\
dependencies:\n\
  - python=3.6\n\
  - pip\n\
  - pip:\n\
      - prophet\n\
' > conda.yaml

RUN apt-get update --yes
# Install gcc
RUN apt-get install build-essential --yes
RUN conda env create --file conda.yaml
```

Reply:
I guess what we really need to install is gcc, not prophet's dependencies.

Reply:
@harupy it passes CI, including serving tests, if these extra dependencies are included, but we don’t seem to be installing gcc in the test environment. That’s the part I still don’t understand - why does it fail in your docker container & succeed in our CI system when gcc isn’t installed in either place?

Reply:
In GitHub Actions, lots of packages are pre-installed:

https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#installed-apt-packages

![image](https://user-images.githubusercontent.com/17039389/132660259-74b83f8d-3f03-44c5-98b2-623cf3955d38.png)

Reply:
Confirmed gcc is installed on github actions:

https://github.com/mlflow/mlflow/pull/4786/checks?check_run_id=3554454258#step:3:958

```
$ which gcc
/usr/bin/gcc
```

Reply:
Testing conda env creation with only including prophet in this PR:
https://github.com/mlflow/mlflow/pull/4786/

Reply:
The following two sets of requirements are equivalent:

```
Cython
convertdate
LunarCalendar
holidays
tqdm
cmdstanpy
setuptools-git
python-dateutil
pystan
prophet
```

```
prophet
```

In either case, pip fails to build a wheel for prophet but manages to install prophet using the legacy `setup.py install` method.

Reply:
Did some testing with removing all of the dependencies (I actually think that the condition that I was dealing with when I saw those issues was due to copying the sklearn log_model declaration in test_pyfunc_serve_and_score. 
Thanks for the 2nd set of eyes on this and validating that the setup.py install is going to be defaulted to regardless of whether we have the build env all set before issuing the prophet pip install. 
Removing those sub dependencies.